lib/sendgrid_toolkit/v3/abstract_sendgrid_client.rb

```diff
@@ -0,0 +1,61 @@
+module SendgridToolkit
+  module V3
+    class AbstractSendgridClient
+
+      BASE_URI = 'https://api.sendgrid.com/v3/asm'
+
+      def initialize(api_user = nil, api_key = nil)
+        @api_user = api_user || SendgridToolkit.api_user || ENV['SMTP_USERNAME']
+        @api_key = api_key || SendgridToolkit.api_key || ENV['SMTP_PASSWORD']
+
+        fail SendgridToolkit::NoAPIUserSpecified if @api_user.nil? || @api_user.length == 0
+        fail SendgridToolkit::NoAPIKeySpecified if @api_key.nil? || @api_key.length == 0
+      end
+
+      protected
+
+      def api_post(action_name, options = {})
+        response = HTTParty.post("#{BASE_URI}/#{action_name}",
+                                 body: options.to_json, format: :json,
+                                 headers: { 'Authorization' => "Basic #{credentials}" })
+        check_response(response)
+      end
+
+      def api_get(action_name, options = {})
+        response = HTTParty.get("#{BASE_URI}/#{action_name}",
+                                query: options, format: :json, headers: { 'Authorization' => "Basic #{credentials}" })
+        check_response(response)
+      end
+
+      def api_delete(action_name, options = {})
+        response = HTTParty.delete("#{BASE_URI}/#{action_name}",
+                                   body: options, format: :json,
+                                   headers: { 'Authorization' => "Basic #{credentials}" })
```

Comment:
@AlexKlim this doesn't seem to work anymore since Ruby `v2.4.6`. It fails because the result header value contains CR/LF.

@freerobby can we consider this as a bug? I tried a freshly generated V3 API key and V3 API request fails with the following error:

```
ArgumentError: header Authorization has field value "Basic ...", this cannot include CR/LF
```

Reply:
@mkalygin Definitely seems like a bug. Candidly, I'm not using this library these days and probably won't have time to fix anytime soon, but would happily take a PR and/or give rights to someone who wants to maintain it going forwad.